src/filter/http_context.rs

```diff
@@ -1,159 +1,259 @@
-
-        let path_bytes = self.get_property(vec!["request", "path"]).unwrap();
-        let path = String::from_utf8(path_bytes).unwrap();
-
-        let method_bytes = self.get_property(vec!["request", "method"]).unwrap();
-        let method = String::from_utf8(method_bytes).unwrap();
-
-        RequestInfo { host, path, method }
+        split_host[0].to_owned()
     }
 
-    fn create_ratelimit_request(
-        &self,
-        domain: &str,
-        actions: &[RLA_action_specifier],
-    ) -> Result<RateLimitRequest, UtilsErr> {
+    fn process_rate_limit_policy(&self, rlp: &RateLimitPolicy) -> Action {
+        let descriptors = self.build_descriptors(rlp);
+        if descriptors.len() == 0 {
+            debug!("[context_id: {}] empty descriptors", self.context_id);
+            return Action::Continue;
+        }
+
         let mut rl_req = RateLimitRequest::new();
+        rl_req.set_domain(rlp.rate_limit_domain.clone());
+        rl_req.set_hits_addend(1);
+        rl_req.set_descriptors(descriptors);
 
-        rl_req.set_domain(domain.into());
+        let rl_req_serialized = Message::write_to_bytes(&rl_req).unwrap(); // TODO(rahulanand16nov): Error Handling
 
-        rl_req.set_hits_addend(1);
+        match self.dispatch_grpc_call(
+            rlp.upstream_cluster.as_str(),
+            RATELIMIT_SERVICE_NAME,
+            RATELIMIT_METHOD_NAME,
+            Vec::new(),
+            Some(&rl_req_serialized),
+            Duration::from_secs(5),
+        ) {
+            Ok(call_id) => info!("Initiated gRPC call (id# {}) to Limitador", call_id),
+            Err(e) => {
+                warn!("gRPC call to Limitador failed! {:?}", e);
+                request_process_failure(self.config.failure_mode_deny);
+            }
+        }
+        Action::Pause
+    }
 
-        rl_req
-            .mut_descriptors()
-            .push(descriptor_from_actions(self, actions)?);
+    fn build_descriptors(
+        &self,
+        rlp: &RateLimitPolicy,
+    ) -> ::protobuf::RepeatedField<RateLimitDescriptor> {
+        //::protobuf::RepeatedField::default()
+        rlp.gateway_actions
+            .iter()
+            .filter(|ga| self.filter_configurations_by_rules(&ga.rules))
+            // filter None and return configurations
+            .flat_map(|ga| &ga.configurations)
+            // flatten the vec<vec<Configurations> to vec<Configuration>
+            .flatten()
+            // filter None and return actions.
+            // All actions cannot be flatten! each vec of actions defines one potential descriptor
+            .flat_map(|configuration| &configuration.actions)
+            // filter None and return descriptors
+            .flat_map(|actions| self.build_single_descriptor(actions))
+            .collect()
+    }
 
-        Ok(rl_req)
+    fn filter_configurations_by_rules(&self, rules_opt: &Option<Vec<Rule>>) -> bool {
+        match rules_opt {
+            // no rules is equivalent to matching all the requests.
```

Comment:
This bits confuses me… (see below)
But relatively the same point of why we warp `Vec`s in `Option`? I feel I'm missing something. Especially that, again if I read this right, `Vec.is_empty() != None`… Or I'm just too tired and reading this all wrong ¯\_(°_o)_/¯